kv/kvserver: TestSchedulerBuffering failed

kv/kvserver.TestSchedulerBuffering [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8108933?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8108933?buildTab=artifacts#/) on master @ [9c5375f6a7375724cdbcbaa0029ed97a230d7abe](https://github.com/cockroachdb/cockroach/commits/9c5375f6a7375724cdbcbaa0029ed97a230d7abe):


```
=== RUN   TestSchedulerBuffering
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestSchedulerBuffering2749501448
    test_log_scope.go:79: use -show-logs to present logs inline
    scheduler_test.go:296: condition failed to evaluate within 45s: expected ready=[1:2] request=[1:2] tick=[1:20], but got ready=[1:2] request=[1:2] tick=[1:24]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestSchedulerBuffering2749501448
--- FAIL: TestSchedulerBuffering (45.13s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSchedulerBuffering.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-22813

kv/kvserver.TestSchedulerBuffering [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8181226?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8181226?buildTab=artifacts#/) on master @ [2c3d75f1ce31024d7ffe530f91f22162c053abcd](https://github.com/cockroachdb/cockroach/commits/2c3d75f1ce31024d7ffe530f91f22162c053abcd):


```
=== RUN   TestSchedulerBuffering
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestSchedulerBuffering2007008249
    test_log_scope.go:79: use -show-logs to present logs inline
    scheduler_test.go:296: condition failed to evaluate within 45s: expected ready=[1:2] request=[1:2] tick=[1:20], but got ready=[1:2] request=[1:2] tick=[1:24]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestSchedulerBuffering2007008249
--- FAIL: TestSchedulerBuffering (45.14s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSchedulerBuffering.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/scheduler_test.go
+++ b/pkg/kv/kvserver/scheduler_test.go
@@ -165,6 +165,7 @@ type testProcessor struct {
 		raftReady   map[roachpb.RangeID]int
 		raftRequest map[roachpb.RangeID]int
 		raftTick    map[roachpb.RangeID]int
+		ready       func()
 	}
 }
 
@@ -176,9 +177,19 @@ func newTestProcessor() *testProcessor {
 	return p
 }
 
+func (p *testProcessor) onReady(f func()) {
+	p.mu.Lock()
+	p.mu.ready = f
+	p.mu.Unlock()
+}
+
 func (p *testProcessor) processReady(rangeID roachpb.RangeID) {
 	p.mu.Lock()
 	p.mu.raftReady[rangeID]++
+	if p.mu.ready != nil {
+		p.mu.ready()
+		p.mu.ready = nil
+	}
 	p.mu.Unlock()
 }
 
@@ -271,6 +282,7 @@ func TestSchedulerBuffering(t *testing.T) {
 
 	testCases := []struct {
 		flag  raftScheduleFlags
+		slow  bool // true if Ready processing is slow
 		ticks int
 		want  string
 	}{
@@ -280,18 +292,43 @@ func TestSchedulerBuffering(t *testing.T) {
 		{flag: stateRaftReady | stateRaftRequest | stateRaftTick,
 			want: "ready=[1:2] request=[1:2] tick=[1:10]"},
 		{flag: stateRaftTick, want: "ready=[1:2] request=[1:2] tick=[1:15]"},
-
-		// Only 5/10 ticks are buffered.
-		{flag: 0, ticks: 10, want: "ready=[1:2] request=[1:2] tick=[1:20]"},
+		// All 4 ticks are processed.
+		{flag: 0, ticks: 4, want: "ready=[1:2] request=[1:2] tick=[1:19]"},
+		// Only 5/10 ticks are buffered while Raft processing is slow.
+		{flag: stateRaftReady, slow: true, ticks: 10, want: "ready=[1:3] request=[1:2] tick=[1:24]"},
+		// All 3 ticks are processed even if processing is slow.
+		{flag: stateRaftReady, slow: true, ticks: 3, want: "ready=[1:4] request=[1:2] tick=[1:27]"},
 	}
 
 	for _, c := range testCases {
+		var started, done chan struct{}
+		if c.slow {
+			started, done = make(chan struct{}), make(chan struct{})
+			p.onReady(func() {
+				close(started)
+				<-done
+			})
+		}
+
 		const id = roachpb.RangeID(1)
-		cnt := s.enqueueN(c.flag, id, id, id, id, id)
+		if c.flag != 0 {
+			s.signal(s.enqueueN(c.flag, id, id, id, id, id))
+			if started != nil {
+				<-started // wait until slow Ready processing has started
+				// NB: this is necessary to work around the race between the subsequent
+				// event enqueue calls and the scheduler loop which may process events
+				// before we've added them. This makes the test deterministic.
+			}
+		}
+
+		cnt := 0
 		for t := c.ticks; t > 0; t-- {
 			cnt += s.enqueue1(stateRaftTick, id)
 		}
 		s.signal(cnt)
+		if done != nil {
+			close(done) // finish slow Ready processing to unblock progress
+		}
 
 		testutils.SucceedsSoon(t, func() error {
 			if s := p.String(); c.want != s {

